cmd/serve/s3/server.go

```diff
@@ -60,17 +73,55 @@ func newServer(ctx context.Context, f fs.Fs, opt *Options) (s *Server, err error
 		gofakes3.WithIntegrityCheck(true), // Check Content-MD5 if supplied
 	)
 
+	w.handler = http.NewServeMux()
+	w.handler = w.faker.Server()
+
+	if proxyflags.Opt.AuthProxy != "" {
+		w.proxy = proxy.New(ctx, &proxyflags.Opt)
+		// proxy auth middleware
+		w.handler = proxyAuthMiddleware(w.handler, w)
+	} else {
+		w._vfs = vfs.New(f, &vfsflags.Opt)
+
+	}
+
 	w.Server, err = httplib.NewServer(ctx,
 		httplib.WithConfig(opt.HTTP),
+		httplib.WithAuth(opt.Auth),
 	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to init server: %w", err)
 	}
 
-	w.handler = w.faker.Server()
 	return w, nil
 }
 
+func (w *Server) getVFS(ctx context.Context) (VFS *vfs.VFS, err error) {
+	if w._vfs != nil {
+		return w._vfs, nil
+	}
+
+	value := ctx.Value(ctxKeyId)
+	if value == nil {
+		return nil, errors.New("no VFS found in context")
+	}
+
+	VFS, ok := value.(*vfs.VFS)
+	if !ok {
+		return nil, fmt.Errorf("context value is not VFS: %#v", value)
+	}
+	return VFS, nil
+}
+
+// auth does proxy authorization
+func (w *Server) auth(accessKeyId string) (value interface{}, err error) {
```

Comment:
To test use this example proxy:
```py
#!/usr/bin/env python3
"""
A proxy for rclone serve s3

This takes the incoming access_key_id
and converts it into an webdav backend with bearer_token auth
"""

import sys
import json

def main():
    i = json.load(sys.stdin)
    o = {
        "type": "webdav",            # type of backend
        "_root": "",                 # root of the fs
        # NOTE: adding user doesn't work with bearer_token (?)
        # "user": i['user'],
        "bearer_token": i['pass'],
        "url": "https://localhost:9200/remote.php/webdav",
    }
    json.dump(o, sys.stdout, indent="\t")

if __name__ == "__main__":
    main()

```

Reply:
@ncw Could you have a look whether the `--auth-proxy` implementation is correct or not?